File: 1307-ugly-number-iii/ugly-number-iii.cpp
class Solution {
public:
    long long gcd(long long a, long long b) {
        if (b == 0) {
            return a;
        }
        return gcd(min(a, b), max(a, b) % min(a, b));
    }

    int nthUglyNumber(int n, int a, int b, int c) {
        long long a_ll = a;
        long long b_ll = b;
        long long c_ll = c;

        long long ab = (a_ll / gcd(a_ll, b_ll)) * b_ll;
        long long ac = (a_ll / gcd(a_ll, c_ll)) * c_ll;
        long long bc = (b_ll / gcd(b_ll, c_ll)) * c_ll;
        
        long long abc = (ab / gcd(ab, c_ll)) * c_ll;

        long long low = 0;
        long long high = 2000000000; 
        long long result = 0;

        while (low <= high) {
            long long mid = low + (high - low) / 2;
            
            long long count = (mid / a_ll) + (mid / b_ll) + (mid / c_ll) 
                           - (mid / ab) - (mid / ac) - (mid / bc)
                           + (mid / abc);

            if (count >= n) {
                result = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return static_cast<int>(result);
    }
};

File: 1307-ugly-number-iii/ugly-number-iii.py
class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        def gcd(a, b):
            if b == 0:
                return a
            return gcd(min(a, b), max(a, b) % min(a, b))
            
        ab = (a * b) // gcd(a, b)
        ac = (a * c) // gcd(a, c)
        bc =  (b * c) // gcd(b, c)
        abc = (ab * c) // gcd(ab, c)
        def check(mid):
            m = (mid // a) + (mid // b) + (mid // c) - (mid // (ab)) - (mid // bc)- (mid // (ac)) + (mid // (abc))
            return n <= m
        low = 0
        high = 2*10**9
        while low + 1 < high:
            mid = (low + high) // 2
            if check(mid): high = mid
            else: low = mid
        return high